Support serialization of fixed-size arrays in abi_serializer.

greg7mdp opened the pull request:
[partially] resolves #1931.

- add support for serialization of fixed size array types such as `uint8[5]`.
- [perf] add missing `reserve()` when deserializing a vector (or now an array), fix unnecessary copy in `vector<fc::variant> vars = var.get_array();`
- add tests (thanks to Kevin) for serialization of fixed size array types.

greg7mdp committed:
wip to support arrays in abi_serializer.

greg7mdp committed:
Fix issue - now `abi_tests.cpp` do pass.

greg7mdp commented:
I think we should check that the length of the array passed to serialize matches the size in the `rtype`.

heifner commented on libraries/chain/abi_serializer.cpp:
```diff
@@ -400,15 +400,15 @@ namespace eosio { namespace chain {
       auto ftype = fundamental_type(rtype);
-      auto btype = built_in_types.find(ftype );
-      if( btype != built_in_types.end() ) {
+      bool var_is_array = is_szarray(rtype);
```
I think this can be:
`bool var_is_array = is_array(rtype) || is_szarray(rtype);`
and then just `var_is_array` below.

greg7mdp replied:
I see your point, but maybe we want a different serialization for fixed-size arrays without the size, as Matt's mentioned in telegram. Let le try that.

heifner commented on libraries/chain/abi_serializer.cpp:
```diff
@@ -496,10 +496,11 @@ namespace eosio { namespace chain {
 
-      auto btype = built_in_types.find(fundamental_type(rtype));
-      if( btype != built_in_types.end() ) {
+      bool var_is_array = is_szarray(rtype);
```
I think this can be:
`bool var_is_array = is_array(rtype) || is_szarray(rtype);`
and then just `var_is_array` below.

heifner commented on libraries/chain/abi_serializer.cpp:
```diff
@@ -400,15 +400,15 @@ namespace eosio { namespace chain {
+      bool var_is_array = is_szarray(rtype);
+      if( auto btype = built_in_types.find(ftype ); !var_is_array && btype != built_in_types.end() ) {
          try {
             return btype->second.first(stream, is_array(rtype), is_optional(rtype), ctx.get_yield_function());
```
I think you can also hard-code `false` for `is_array(rtype)`

heifner commented on libraries/chain/abi_serializer.cpp:
```diff
@@ -496,10 +496,11 @@ namespace eosio { namespace chain {
+      bool var_is_array = is_szarray(rtype);
+      
+      if( auto btype = built_in_types.find(fundamental_type(rtype)); !var_is_array && btype != built_in_types.end() ) {
          btype->second.second(var, ds, is_array(rtype), is_optional(rtype), ctx.get_yield_function());
```
I think you can also hard-code `false` for `is_array(rtype)`

heifner commented on unittests/abi_tests.cpp:
```diff
@@ -530,6 +530,44 @@ fc::variant verify_type_round_trip_conversion( const abi_serializer& abis, const
+BOOST_AUTO_TEST_CASE(std_array_types)
+{ try {
+
+   const char* currency_abi = R"=====(
```
Rename to `test_abi`

greg7mdp replied:
I don't really mind, but this doesn't seem consistent with the other tests in the same file.

heifner replied:
I mean `currency_abi` => `test_abi`

greg7mdp replied:
Got it, makes sense, thanks!

greg7mdp committed:
Better support for fixed-size arrays (don't serialize the size)

heifner commented on libraries/chain/abi_serializer.cpp:
```diff
@@ -218,17 +218,23 @@ namespace eosio { namespace chain {
    }
 
-   bool abi_serializer::is_szarray(const string_view& type)const {
+   std::optional<fc::unsigned_int> abi_serializer::is_szarray(const string_view& type)const {
```
I think we need a test that demonstrates unmatched `std::array` size does not work and generates a meaningful error.

greg7mdp replied:
done.

greg7mdp committed:
Update abi name in test: `currency_abi` -> `test_abi`

greg7mdp committed:
Add test where fixed-size array is serialized with wrong number of values.

Also replace the mouthful `abi_serializer::create_yield_function( max_serialization_time )` with `yield_fn()`

greg7mdp committed:
Merge branch 'main' of github.com:AntelopeIO/leap into abi_serializer_array

greg7mdp commented:
Replaced by https://github.com/AntelopeIO/spring/pull/1614
